dj/construction/extract.py

```diff
@@ -0,0 +1,541 @@
+    for namespaces_cols in namespaces.values():
+        for col in namespaces_cols:  # type: ignore
+            if col in no_namespace_safe_cols:
+                multiple_refs.add(col)
+            no_namespace_safe_cols.add(col)
+
+    namespaces[""] = no_namespace_safe_cols - multiple_refs  # type: ignore
+
+    def check_col(col: Column, add: list) -> Optional[str]:
+        """Check if a column can be had in the query"""
+        namespace = make_name(col.namespace)  # str preceding the column name
+        cols = namespaces.get(namespace)
+        if (
+            cols is None
+        ):  # there is just no namespace at all where the node could come from
+            CompoundBuildException().catch(
+                error=DJError(
+                    code=ErrorCode.MISSING_COLUMNS,
+                    message=(
+                        f"No namespace `{namespace}` from which to reference column "
+                        f"`{col.name.name}`."
+                    ),
+                    context=str(col.parent),
+                ),
+                message="Cannot extract dependencies from SELECT.",
+            )
+            return None
+        if (
+            col.name.name not in cols
+        ):  # the proposed namespace does not contain the column; which error to raise?
+            exc_msg = f"Namespace `{namespace}` has no column `{col.name.name}`."
+            error_code = ErrorCode.MISSING_COLUMNS
+            if not namespace:
+                exc_msg = (
+                    f"Column `{col.name.name}` does not exist in any referenced tables."
+                )
+                if col.name.name in multiple_refs:
+                    exc_msg = f"`{col.name.name}` appears in multiple references and so must be namespaced."  # pylint: disable=C0301
+                    error_code = ErrorCode.INVALID_SQL_QUERY
+            CompoundBuildException().catch(
+                error=DJError(
+                    code=error_code,
+                    message=exc_msg,
+                    context=str(col.parent),
+                ),
+                message="Cannot extract dependencies from SELECT",
+            )
+
+            return None
+        if namespace:  # there is a proposed namespace that has the column
+            add.append((col, table_nodes[namespace]))
+        else:  # finally check if the column that does not have a namespace is in any namespace
+            for nmpsc, nmspc_cols in namespaces.items():  # pragma: no cover
+                if col.name.name in nmspc_cols:
+                    add.append((col, table_nodes[nmpsc]))
+                    return namespace
+        return namespace
+
+    # check projection
+    for col in chain(*(exp.find_all(Column) for exp in select.projection)):
+        check_col(col, table_deps.columns.projection)  # type: ignore
+
+    # check groupby, filters, and join ons
+    gbfo: List[
+        Tuple[List[Tuple[Column, Union[TableExpression, Node]]], Column, bool]
```

Comment:
Yeah, it's a poorly named attempt to organize the variables I wanted for the loop here https://github.com/DataJunction/dj/blob/619987c49293a38cdf1cdc6169c3070ebb429201/dj/construction/extract.py#L395

Breaking down that `Tuple`

- `List[Tuple[Column, Union[TableExpression, Node]]]`: this will be what we append our pairs of resolved `Column`, `Node` to (Column here are AST nodes and Node is DJ Node 😵‍💫)... TableExpression should only be subquery so `Alias[Select]` or `Select` here so I might tighten that type up
- `Column`: This is the current column we are looking at and if it checks out we will append it as mentioned above
- `bool`: ugly imo, but I'm using this to differentiate between where a dimension would be referenced or not... This `extract_dependencies_from_select` design allows you to take an AST modified to include dimensions/filters referenced from the API and process them as if there was SQL submitted that directly referenced dimension nodes in a GROUP BY, WHERE, HAVING, or any JOIN's ON. Ofc the dimension would not be valid for the JOIN though so this `bool` is basically to indicate the column came from an ON

Reply:
I see. Yeah this is not easy to remember but I don't have a better suggestion here. We can always clean it up later...